go/vt/vtgate/evalengine/weights.go

```diff
@@ -0,0 +1,175 @@
+		return nil, true, nil
+
+	case sqltypes.IsSigned(coerceTo):
+		i, err := v.ToInt64()
+		if err != nil {
+			return dst, false, err
+		}
+		raw := uint64(i)
+		raw = raw ^ (1 << 63)
+		return binary.BigEndian.AppendUint64(dst, raw), true, nil
+
+	case sqltypes.IsUnsigned(coerceTo):
+		u, err := v.ToUint64()
+		if err != nil {
+			return dst, false, err
+		}
+		return binary.BigEndian.AppendUint64(dst, u), true, nil
+
+	case sqltypes.IsFloat(coerceTo):
+		f, err := v.ToFloat64()
+		if err != nil {
+			return dst, false, err
+		}
+
+		raw := math.Float64bits(f)
+		if math.Signbit(f) {
+			raw = ^raw
+		} else {
+			raw = raw ^ (1 << 63)
+		}
+		return binary.BigEndian.AppendUint64(dst, raw), true, nil
+
+	case sqltypes.IsBinary(coerceTo):
+		b := v.Raw()
+		if length != 0 {
+			if length > cap(b) {
+				b = append(b, make([]byte, length-len(b))...)
+			} else {
+				b = b[:length]
+			}
+		}
+		return append(dst, b...), false, nil
+
+	case sqltypes.IsText(coerceTo):
+		coll := col.Get()
+		if coll == nil {
+			panic("cannot hash unsupported collation")
+		}
+		b := v.Raw()
+		if length != 0 {
+			b = charset.Slice(coll.Charset(), b, 0, length)
+		}
+		return coll.WeightString(dst, b, length), false, nil
+
+	case sqltypes.IsDecimal(coerceTo):
+		dec, err := decimal.NewFromMySQL(v.Raw())
+		if err != nil {
+			return dst, false, err
+		}
+		return dec.WeightString(dst, int32(length), int32(precision)), true, nil
+	case coerceTo == sqltypes.TypeJSON:
+		j, err := json.NewFromSQL(v)
+		if err != nil {
+			return dst, false, err
+		}
+		return j.WeightString(dst), false, nil
+	default:
+		return fallbackWeightString(dst, v, coerceTo, col, length, precision)
+	}
+}
+
+func fallbackWeightString(dst []byte, v sqltypes.Value, coerceTo sqltypes.Type, col collations.ID, length, precision int) ([]byte, bool, error) {
+	e, err := valueToEvalCast(v, coerceTo, col)
+	if err != nil {
+		return dst, false, err
+	}
+	return evalWeightString(dst, e, length, precision)
+}
+
+func evalWeightString(dst []byte, e eval, length, precision int) ([]byte, bool, error) {
+	switch e := e.(type) {
+	case nil:
+		return nil, true, nil
+	case *evalInt64:
+		raw := uint64(e.i)
+		raw = raw ^ (1 << 63)
+		return binary.BigEndian.AppendUint64(dst, raw), true, nil
+	case *evalUint64:
+		return binary.BigEndian.AppendUint64(dst, e.u), true, nil
+	case *evalFloat:
+		raw := math.Float64bits(e.f)
+		if math.Signbit(e.f) {
+			raw = ^raw
+		} else {
+			raw = raw ^ (1 << 63)
+		}
+		return binary.BigEndian.AppendUint64(dst, raw), true, nil
+	case *evalDecimal:
+		return e.dec.WeightString(dst, int32(length), int32(precision)), true, nil
+	case *evalBytes:
+		if e.isBinary() {
+			b := e.bytes
+			if length != 0 {
+				if length > cap(b) {
+					b = append(b, make([]byte, length-len(b))...)
+				} else {
+					b = b[:length]
+				}
+			}
+			return append(dst, b...), false, nil
+		}
+		coll := e.col.Collation.Get()
+		if coll == nil {
+			panic("cannot hash unsupported collation")
+		}
```

Comment:
same here, but i don't have strong opinions. just thought we're already returning an error in this function so we might as well stick to this